### PR TITLE
Fix `uncessary-coding-comment` fix when there's leading content

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/astral-sh/ruff/issues/6756

Including whitespace, code, and continuations.

### CODE REVIEW

File: crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ pub(crate) fn unnecessary_coding_comment(
             let mut diagnostic = Diagnostic::new(UTF8EncodingDeclaration, *range);
             if settings.rules.should_fix(diagnostic.kind.rule()) {
                 diagnostic.set_fix(Fix::automatic(Edit::deletion(
-                    range.start(),
+                    locator.line_start(range.start()),
                     locator.full_line_end(range.end()),
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:71]: What happens if the coding comment is preceded by content, like:

```python
print(x) # coding=utf8
print("Hello world")
```

Comment (zanieb) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:71]: That shouldn't match our regex, but I can add test coverage

Comment (zanieb) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:71]: Except it does get removed.. humph will investigate

Comment (charliermarsh) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:71]: The regular expression is only run over the comment (the range is that of the comment).

Comment (zanieb) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:71]: Yeah I just realized that — this was less clear when the variable was named `line`, pretty obvious now

Comment (zanieb) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs]: [PEP 263](https://peps.python.org/pep-0263/) doesn't let you declare these as comments with other content on the line so I've updated the regex to apply to the full line.

File: crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs [hunk 2]
```diff
@@ -55,20 +55,21 @@ pub(crate) fn unnecessary_coding_comment(
...
             }
 
-            let mut diagnostic = Diagnostic::new(UTF8EncodingDeclaration, *range);
+            let mut diagnostic = Diagnostic::new(UTF8EncodingDeclaration, *comment_range);
             if settings.rules.should_fix(diagnostic.kind.rule()) {
                 diagnostic.set_fix(Fix::automatic(Edit::deletion(
-                    range.start(),
-                    locator.full_line_end(range.end()),
+                    line_range.start(),
+                    line_range.end(),
```
[the remark this answers is not in the document]
Comment (zanieb) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:72]: Yeesh. Alright will fix this.

[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:72]: If we don't, it will get fuzzed eventually :joy:

**Review by charliermarsh — APPROVED** [review]
I think another valid solution here would be: if the comment isn't at the beginning of the line, delete _just_ the comment (not the `locator.full_line_end` of the comment). Possibly a bit simpler since you wouldn't need to expand the line for the regex, wouldn't need to special-case continuations, etc. But defer to you as the author.

**Comment by zanieb** [comment]
Hm that seems kind of nice but in the end we probably don't want to retain that extra content. Are there other rules that would cover the removing the remaining whitespace and newlines? I'm tempted to just move on until we see another edge-case in the future.

> wouldn't need to expand the line for the regex

I still think we should be doing this anyway as the regex is designed to match a full line.

**Comment by charliermarsh** [comment]
I worry that the regex will open the door to other problems though. How does the regex fare with an example like this?

```
"""
# coding=utf8""" # empty comment
```

**Comment by zanieb** [comment]
:sigh: yeah that is broken too. We could address that in the current implementation by making the end of the regex more strict or by applying the regex twice (once to the comment itself and once to the line), but you've convinced me to reconsider the approach.

I'm not sure about just narrowing the match though. Per the PEP

> There must not be any Python statement on the line that contains the encoding declaration. 

we should probably just ignore cases entirely if the line doesn't match the regex. It feels incorrect to delete a comment when it's not a valid coding comment in the first place.

**Comment by charliermarsh** [comment]
What about something like:

- If there is any non-whitespace content between the start of the line and the comment, don't flag it.
- If there is whitespace at the start of the line, only remove the comment range (not the entire line) _or_ expand to include the continuation in the deletion.
- Otherwise, remove the entire line.

**Comment by zanieb** [comment]
I think applying the regex twice is the simplest way to ensure we're doing the right thing here without rewriting this whole thing to use a tokenizer? I could do that here but I'm leaning towards opening that as a follow-up issue for a contributor and merging this fix.

File: crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs [hunk 3]
```diff
@@ -55,20 +55,40 @@ pub(crate) fn unnecessary_coding_comment(
...
-        if CODING_COMMENT_REGEX.is_match(line) {
+    for comment_range in indexer.comment_ranges().iter().take(2) {
+        let line_range = locator.full_line_range(comment_range.start());
+        let line = locator.slice(line_range);
+        let comment = locator.slice(*comment_range);
+
+        // Both the line and the comment itself must match to prevent false positives
+        // where the comment is on a line that includes content that looks like a coding comment
+        // or the coding comment is preceded by non-whitespace
+        if CODING_COMMENT_REGEX.is_match(line) && CODING_COMMENT_REGEX.is_match(comment) {
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pyupgrade/rules/unnecessary_coding_comment.rs:66]: I think I would still recommend doing something like:

```rust
// Check if there's any content preceding the comment.
let line_start = locator.line_start(*range.start());
if !locator
    .slice(TextRange::new(line_start, range.start()))
    .trim()
    .is_empty()
{
    continue;
}
```

Personally, I find that much easier to understand than running the regex twice.

**Review by charliermarsh — APPROVED** [review]

**Comment by zanieb** [comment]
Thanks for the example — I thought you were suggesting using the tokenizer but that's nice and simple.

I think the only thing that differs now is

> If there is whitespace at the start of the line, only remove the comment range (not the entire line) or expand to include the continuation in the deletion.

Comments following continuations are not valid coding definitions as far as I can tell, we probably shouldn't flag them at all. I think it's kind of academic at this point though.

**Review by charliermarsh — COMMENTED** [review]
This looks great!

**Pull request merged**